charms/kfp-profile-controller/tests/integration/test_charm.py

```diff
@@ -35,19 +44,181 @@ async def test_build_and_deploy(ops_test: OpsTest):
+    )
+
     # Maybe: await ops_test.model.wait_for_idle(raise_on_error=False, raise_on_blocked=True) ?
-    await ops_test.model.wait_for_idle(timeout=60 * 10)
-
-
-# TODO: Real testing requires:
-#  * kubeflow-profile controller, or something that will make namespaces/service accounts in a
-#    similar fashion (some resources deployed to each namespace by kfp-profile-controller use the
-#    `default-editor` service account, which is created by the main kubeflow-profile controller)
-#  * This charm deploying successfully is not a full test of function.  We need to create a new
-#    tracked namespace (namespace that has label pipelines.kubeflow.org/enabled=true) and ensure
-#    all expected resources are deployed and come up successfully (eg: if service account is not
-#    found, deployments will exist but pods will never be "ready")
-#  * check whether secrets and other things get updated based on the config sent to sync.py/kfppc
-#  * could create namespaces with the right pipelines label to trigger the sync, but also need to
-#    create a dummy poddefaults CRD too
+    await ops_test.model.wait_for_idle(status="active", timeout=60 * 10)
+
+
+async def test_status(ops_test: OpsTest):
+    charms = ("kubeflow-profiles", "metacontroller-operator", "kfp-profile-controller")
+    for charm in charms:
+        assert ops_test.model.applications[charm].units[0].workload_status == "active"
+
+
+async def test_profile_creation(lightkube_client, profile):
+    # Test whether a namespace was created for this profile
+    profile_name = profile
+    validate_profile_resources(lightkube_client, profile_name)
+
+
+@pytest.fixture(scope="session")
+def lightkube_client() -> lightkube.Client:
+    client = lightkube.Client()
+    create_global_resource(group="kubeflow.org", version="v1", kind="Profile", plural="profiles")
+    return client
+
+
+def _safe_load_file_to_text(filename: str):
+    """Returns the contents of filename if it is an existing file, else it returns filename."""
+    try:
+        text = Path(filename).read_text()
+    except FileNotFoundError:
+        text = filename
+    return text
+
+
+@pytest.fixture(scope="session")
+def profile(lightkube_client):
+    """Creates a Profile object in cluster, cleaning it up after tests."""
+    profile_file = "./tests/integration/profile.yaml"
+    yaml_text = _safe_load_file_to_text(profile_file)
+    yaml_rendered = yaml.safe_load(yaml_text)
+    profilename = yaml_rendered["metadata"]["name"]
+
+    create_all_from_yaml(yaml_file=yaml_text, lightkube_client=lightkube_client)
+    yield profilename
+
+    delete_all_from_yaml(yaml_text, lightkube_client)
+
+
+ALLOWED_IF_EXISTS = (None, "replace", "patch")
+
+
+def _validate_if_exists(if_exists):
+    if if_exists in ALLOWED_IF_EXISTS:
+        return if_exists
+    else:
+        raise ValueError(
+            f"Invalid value for if_exists '{if_exists}'.  Must be one of {ALLOWED_IF_EXISTS}"
+        )
+
+
+def create_all_from_yaml(
```

Comment:
I know who to blame for this function (I believe you pulled it from a past charm of mine :D ), but now that `lightkube` provides an [`apply` function](https://lightkube.readthedocs.io/en/latest/client/) I think this could just use something native from `lightkube`.  Give it a thought and see if you can simplify this

Reply:
I'm not sure I can use the `apply` function. According to lightkube documentation, `server-side apply` requires k8s v1.22 while we're using 1.21. Please correct me if I'm wrong.

Reply:
Based on discussion in our daily, I think this is unblocked now and it `apply` should work